### PR TITLE
Add cohort_visited_no_bp to cohort report

### PR DESCRIPTION
**Pull request by PriyangaPKini**
**Story card:** [sc-6499](https://app.shortcut.com/simpledotorg/story/6499/cohort-break-ups-don-t-add-up-to-100)

## Because

Cohort breakups don't add up to 100%

## This addresses

- Add `cohort_visited_no_bp` to cohort report
- Earlier `no_bp` was calling `cohort_missed_visit` instead of `cohort_visited_no_bp`
- Add controlled_rate, uncontrolled_rate, no_bp_rate and missed_visits_rate to cohort_data hash
- Add rounded_percentage algorithm to round the percentages in cohort report

### CODE REVIEW

**Review by govindkrjoshi — COMMENTED** [review]
Looks good to me

**Review by ssrihari — CHANGES_REQUESTED** [review]
The larger change from the user's PoV is to add the `visited_no_bp` into this graph. Consider changing the story / PR title to that, and moving the rounding change as an additional enhancement.

File: app/services/cohort_service.rb [hunk 1]
```diff
@@ -47,16 +49,30 @@ def compute(results)
       end
       arry << {
```
[the remark this answers is not in the document]
Comment (PriyangaPKini) [reply app/services/cohort_service.rb:50]: It's just a variable name. I will change it to some meaning full name. (`arry` was named by the previous developer)

File: app/models/reports/percentage.rb [hunk 2]
```diff
@@ -9,6 +9,8 @@ def percentage(numerator, denominator, with_rounding: true)
   end
 
   def rounded_percentages(counts_hash)
+    return counts_hash if counts_hash.values.uniq == [nil]
```
[the remark this answers is not in the document]
Comment (PriyangaPKini) [reply app/models/reports/percentage.rb:12]: yeah, will change it

File: app/services/cohort_service.rb [hunk 3]
```diff
@@ -47,16 +49,30 @@ def compute(results)
...
         patients_registered: registration_period.to_s,
         registered: result.cohort_patients,
-        results_in: result.period.to_s(:cohort),
-        uncontrolled: result.cohort_uncontrolled
+        results_in: result.period.to_s(:cohort)
       }.with_indifferent_access
     end
   end
 
+  def cohort_reports_rates(cohort)
```
[the remark this answers is not in the document]
Comment (PriyangaPKini) [reply app/services/cohort_service.rb:67]: right, will do that

File: app/models/reports/percentage.rb [hunk 4]
```diff
@@ -8,7 +9,9 @@ def percentage(numerator, denominator, with_rounding: true)
     end
   end
 
-  def rounded_percentages(counts_hash)
+  memoize def rounded_percentages(counts_hash)
```
Comment (ssrihari) [inline app/models/reports/percentage.rb:12]: Since this is now memoized, should `appts_scheduled_rates` not be memoized any more?

**Review by ssrihari — APPROVED** [review]
One comment, but looks good to merge.

File: app/models/reports/percentage.rb (hunk 4 above)
Comment (PriyangaPKini) [reply app/models/reports/percentage.rb:12]: That's right

**Pull request merged**